Add docker support for web server

halis opened the pull request:
Add scripts for managing docker container

halis committed:
Add docker support for web server

Dockerfile for building the image
Add scripts for managing docker container
Documentation

mvolkmann commented on Dockerfile:
```diff
@@ -0,0 +1,19 @@
+COPY yarn.lock yarn.lock
+RUN yarn
+
+COPY . .
```
Does `COPY . .` also copy `package.json` and `yarn.lock`?
Maybe the two `COPY` lines above this aren't needed.

halis replied:
See https://blog.risingstack.com/minimal-docker-containers-for-node-js/

"So far so good - but what’s going on here? We just copy our package.json into the image, then run npm install then copy the rest. These copy steps are done separately because Docker works using layers that can be cached. So if our package.json does not change, Docker doesn’t need to install the dependencies on each build saving us a lot of time."

mvolkmann replied:
Great! Maybe add a comment about this in `Dockerfile` so me and others won't ask this same question again in the future. ;-)

halis replied:
👍

mvolkmann commented on Dockerfile:
```diff
@@ -0,0 +1,19 @@
+RUN yarn run build
+
+EXPOSE 3000
+ENTRYPOINT ["npm", "start"]
```
Will this run from the files in the `build` directory instead of the original files? I didn't see where it is configured to do that.

halis replied:
Not sure I understand the question.

halis replied:
Basically it copies the package.json and yarn.lock and installs
Then it copies everything else so you can run a build command
Which will generate build and npm start gets run from that directory which contains build

mvolkmann replied:
When we run the UI locally now, it does not use files in the build directory. I just want to make sure when we run the UI in the Docker container, it will run differently and really use the files in the build directory instead of files in the src directory.

halis committed:
Documenation

halis committed:
Express server for production

Serves up static assets under `/build`
Since it’s a single page app, any other route will serve up `/build/index.html`

halis committed:
In docker container use express for production mode

halis committed:
Merge branch 'master' of https://github.com/mvolkmann/starter into feature/docker

* 'master' of https://github.com/mvolkmann/starter:
  added use of Bootstrap
  added tests

# Conflicts:
#	package.json
#	yarn.lock

mvolkmann commented on server.js:
```diff
@@ -0,0 +1,30 @@
+app.use(compression());
+app.use(express.static('build'));
+
+singlePageApp();
```
I'm curious what benefit you see in creating the `singlePageApp` and `start` functions given that each is only called once and each contains a single function call. I'd just move that code to here and remove the function definitions. Otherwise 👍

halis committed:
Make this file valid JSON

The default config I copied had dangling commas which JSON just hates…
